packages/aws-cdk/test/integ/cli/resource-pool.ts

```diff
@@ -0,0 +1,76 @@
+/**
+ * A class that holds a pool of resources and gives them out and returns them on-demand
+ *
+ * The resources will be given out front to back, when they are returned
+ * the most recently returned version will be given out again (for best
+ * cache coherency).
+ *
+ * If there are multiple consumers waiting for a resource, consumers are serviced
+ * in FIFO order for most fairness.
+ */
+export class ResourcePool<A> {
+  private readonly resources: A[];
+  private readonly waiters: Array<(x: A) => void> = [];
+
+  constructor(resources: A[]) {
+    if (resources.length === 0) {
+      throw new Error('Must have at least one resource in the pool');
+    }
+    this.resources = [...resources];
```

Comment:
Dedup maybe? To avoid any funky unexpected repititions

Reply:
Not sure what you mean here. I'm making a copy so we don't have accidental sharing of arrays.

Reply:
I mean that `AWS_REGIONS` might be configured as `['us-east-1', 'us-east-1', 'us-east-1']`, effectively causing three tests to run in parallel on `us-east-1`.

Reply:
Hm. Seems that's the caller's responsibility. Maybe they want to weight certain regions differently? I could see that...

Reply:
Agree as far as this generic pool is concerned. 

But if we want to make sure tests don't run in parallel in the same region then I would add the deduplication on the caller side when you read the `AWS_REGIONS` env variable. Maybe even throw if `len(array) != len(set)`.

I actually don't mind not deduping, just mentioning it because there is an edge case here that might break the intended functionality.